processing/src/main/java/org/apache/druid/query/operator/GlueingPartitioningOperator.java

```diff
@@ -0,0 +1,261 @@
+
+      if (cont.iter != null) {
+        while (cont.iter.hasNext()) {
+          RowsAndColumns next = cont.iter.next();
+
+          if (!cont.iter.hasNext()) {
+            // We are at the last RAC. Process it only if subContinuation is null, otherwise save it in previousRac.
+            if (cont.subContinuation == null) {
+              receiver.push(next);
+              receiver.completed();
+              return null;
+            } else {
+              previousRac = next;
+              break;
+            }
+          }
+
+          final Signal signal = receiver.push(next);
+          switch (signal) {
+            case GO:
+              break;
+            case PAUSE:
+              if (cont.iter.hasNext()) {
+                return cont;
+              }
+
+              if (cont.subContinuation == null) {
+                // We were finished anyway
+                receiver.completed();
+                return null;
+              }
+
+              return new Continuation(null, cont.subContinuation);
+            case STOP:
+              receiver.completed();
+              try {
+                cont.close();
+              }
+              catch (IOException e) {
+                throw new RE(e, "Unable to close continuation");
+              }
+              return null;
+            default:
+              throw new RE("Unknown signal[%s]", signal);
+          }
+        }
+
+        if (cont.subContinuation == null) {
+          receiver.completed();
+          return null;
+        }
+      }
+
+      continuation = cont.subContinuation;
+    }
+
+    AtomicReference<Iterator<RowsAndColumns>> iterHolder = new AtomicReference<>();
+
+    final Closeable retVal = child.goOrContinue(
+        continuation,
+        new Receiver()
+        {
+          @Override
+          public Signal push(RowsAndColumns rac)
+          {
+            if (rac == null) {
+              throw DruidException.defensive("Should never get a null rac here.");
+            }
+
+            ClusteredGroupPartitioner groupPartitioner = rac.as(ClusteredGroupPartitioner.class);
+            if (groupPartitioner == null) {
+              groupPartitioner = new DefaultClusteredGroupPartitioner(rac);
+            }
+
+            int[] computedBoundaries = groupPartitioner.computeBoundaries(partitionColumns);
+
+            final ArrayList<RowsAndColumns> gluedRACs = getGluedRACs(rac, computedBoundaries);
+            Iterator<RowsAndColumns> partitionsIter = gluedRACs.iterator();
+
+            Signal keepItGoing = Signal.GO;
+            while (keepItGoing == Signal.GO && partitionsIter.hasNext()) {
+              RowsAndColumns rowsAndColumns = partitionsIter.next();
+              if (partitionsIter.hasNext()) {
+                keepItGoing = receiver.push(rowsAndColumns);
+              } else {
+                // If it's the last element, save it in previousRac instead of pushing to receiver.
+                previousRac = rowsAndColumns;
```

Comment:
I find the current contracts quite odd; 

* this `Reciever`  should be a static inner class (there's no need to create it on every invocation ; and I guess it will be cleaner)
* the `previousRac` should be a field in that class
* I think the `getGluedRACs` method should should not exists as it is now; instead it should be an iterator which may do the glueing (or not) on the 1st returned partition if it exists depending on the `previousRac` value
* also the `previousRac` should be cleared right away if `next` returns; as if the iterator's `next` is consumed it will be sent forward
   * it seems to me that if the `receiver` returns `!GO` the old value will be retained which could cause duplicate results to appear and cause incorrect results
   * cover this with a testcase?

Reply:
> this Reciever should be a static inner class (there's no need to create it on every invocation ; and I guess it will be cleaner)

Do you mean having a `static class GlueingReceiver implements Receiver {}`, having a class variable `private static GlueingReceiver glueingReceiver;`, initializing it in the GlueingPartitioningOperator's constructor, and using the same `glueingReceiver` everywhere in the class?

> the previousRac should be a field in that class

But we also need previousRac to be accessible in the continuation if-else flows. So would you suggest having getter/setter for it in the static class?

> I think the getGluedRACs method should should not exists as it is now; instead it should be an iterator which may do the glueing (or not) on the 1st returned partition if it exists depending on the previousRac value

Can you elaborate on this suggestion? We are creating an iterator from the returned ArrayList. Do you mean that we don't need to create the ArrayList at all?

> also the previousRac should be cleared right away if next returns; as if the iterator's next is consumed it will be sent forward. it seems to me that if the receiver returns !GO the old value will be retained which could cause duplicate results to appear and cause incorrect results. cover this with a testcase?

Will think about this and try adding a test case 🤔

Reply:
> it seems to me that if the receiver returns !GO the old value will be retained which could cause duplicate results to appear and cause incorrect results

Can you describe this scenario with some example?

In the pause-resume scenario, we are handling the last element in the following code block:
```java
if (!cont.iter.hasNext()) {
            // We are at the last RAC. Process it only if subContinuation is null, otherwise save it in previousRac.
            if (cont.subContinuation == null) {
              receiver.push(next);
              receiver.completed();
              return null;
            } else {
              previousRac = next;
              break;
            }
          }
```

What else needs to be handled? 🤔 

Also, all the tests call `runToCompletion()` which internally does `run without pause + run with pauses`, so pause scenario tests are covered there already?

Reply:
> I think the getGluedRACs method should should not exists as it is now; instead it should be an iterator which may do the glueing (or not) on the 1st returned partition if it exists depending on the previousRac value

I've made the change to create a class `GluedRACsIterator implements Iterator<RowsAndColumns>`, and encapsulated the glueing logic in the `next()` method. Please let me know if this is what you had in mind. Thanks!

Reply:
Based on offline discussion:

> this Reciever should be a static inner class (there's no need to create it on every invocation ; and I guess it will be cleaner). the previousRac should be a field in that class

This isn't needed right now, can be done later. Right now we can/should just focus on moving the redundant code to a base partitioning class.

> I think the getGluedRACs method should should not exists as it is now; instead it should be an iterator which may do the glueing (or not) on the 1st returned partition if it exists depending on the previousRac value

This is done in the latest PR code.

> also the previousRac should be cleared right away if next returns; as if the iterator's next is consumed it will be sent forward

This is already handled in the PR code.